// cockatiel/dist/BulkheadPolicy.js
import { neverAbortedSignal } from './common/abort.js';
import { defer } from './common/defer.js';
import { EventEmitter } from './common/Event.js';
import { ExecuteWrapper } from './common/Executor.js';
import { BulkheadRejectedError } from './errors/BulkheadRejectedError.js';
import { TaskCancelledError } from './errors/Errors.js';
export class BulkheadPolicy {
    capacity;
    queueCapacity;
    active = 0;
    queue = [];
    onRejectEmitter = new EventEmitter();
    executor = new ExecuteWrapper();
    /**
     * @inheritdoc
     */
    onSuccess = this.executor.onSuccess;
    /**
     * @inheritdoc
     */
    onFailure = this.executor.onFailure;
    /**
     * Emitter that fires when an item is rejected from the bulkhead.
     */
    onReject = this.onRejectEmitter.addListener;
    /**
     * Returns the number of available execution slots at this point in time.
     */
    get executionSlots() {
        return this.capacity - this.active;
    }
    /**
     * Returns the number of queue slots at this point in time.
     */
    get queueSlots() {
        return this.queueCapacity - this.queue.length;
    }
    /**
     * Bulkhead limits concurrent requests made.
     */
    constructor(capacity, queueCapacity) {
        this.capacity = capacity;
        this.queueCapacity = queueCapacity;
    }
    /**
     * Executes the given function.
     * @param fn Function to execute
     * @throws a {@link BulkheadRejectedException} if the bulkhead limits are exceeeded
     */
    async execute(fn, signal = neverAbortedSignal) {
        if (signal.aborted) {
            throw new TaskCancelledError();
        }
        if (this.active < this.capacity) {
            this.active++;
            try {
                return await fn({ signal });
            }
            finally {
                this.active--;
                this.dequeue();
            }
        }
        if (this.queue.length < this.queueCapacity) {
            const { resolve, reject, promise } = defer();
            this.queue.push({ signal, fn, resolve, reject });
            return promise;
        }
        this.onRejectEmitter.emit();
        throw new BulkheadRejectedError(this.capacity, this.queueCapacity);
    }
    dequeue() {
        const item = this.queue.shift();
        if (!item) {
            return;
        }
        Promise.resolve().then(() => {
            if (item.signal.aborted) {
                item.reject(new TaskCancelledError());
                this.dequeue();
                return;
            }
            this.execute(item.fn, item.signal).then(item.resolve, item.reject);
        });
    }
}
